File: packages/flet/lib/src/controls/cupertino_textfield.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import '../extensions/control.dart';
import '../models/control.dart';
import '../utils/autofill.dart';
import '../utils/borders.dart';
import '../utils/box.dart';
import '../utils/colors.dart';
import '../utils/edge_insets.dart';
import '../utils/form_field.dart';
import '../utils/gradient.dart';
import '../utils/images.dart';
import '../utils/layout.dart';
import '../utils/misc.dart';
import '../utils/numbers.dart';
import '../utils/platform.dart';
import '../utils/text.dart';
import '../utils/textfield.dart';
import '../utils/theme.dart';
import 'base_controls.dart';

class CupertinoTextFieldControl extends StatefulWidget {
  final Control control;

  CupertinoTextFieldControl({Key? key, required this.control})
      : super(key: key ?? ValueKey("control_${control.id}"));

  @override
  State<CupertinoTextFieldControl> createState() =>
      _CupertinoTextFieldControlState();
}

class _CupertinoTextFieldControlState extends State<CupertinoTextFieldControl> {
  String _value = "";
  bool _focused = false;
  bool _revealPassword = false;
  late TextEditingController _controller;
  late final FocusNode _focusNode;
  late final FocusNode _shiftEnterfocusNode;
  String? _lastFocusValue;
  String? _lastBlurValue;
  TextSelection? _selection;

  KeyEventResult _handleTextFieldKeyEvent(KeyEvent event,
      {required bool submitOnEnter}) {
    // ignore up/down arrow keys if flag is set
    if ((event is KeyDownEvent || event is KeyRepeatEvent) &&
        widget.control.getBool("ignore_up_down_keys", false)! &&
        (event.logicalKey == LogicalKeyboardKey.arrowUp ||
            event.logicalKey == LogicalKeyboardKey.arrowDown)) {
      return KeyEventResult.handled;
    }

    // submit on Enter if flag is set and shift is not pressed
    if (submitOnEnter &&
        event is KeyDownEvent &&
        !HardwareKeyboard.instance.isShiftPressed &&
        (event.logicalKey == LogicalKeyboardKey.enter ||
            event.logicalKey == LogicalKeyboardKey.numpadEnter)) {
      widget.control.triggerEvent("submit");
      return KeyEventResult.handled;
    }

    // let the system handle other key events
    return KeyEventResult.ignored;
  }

  @override
  void initState() {
    super.initState();
    _controller = TextEditingController();
    _controller.addListener(_handleControllerChange);
    _shiftEnterfocusNode = FocusNode(
      onKeyEvent: (FocusNode node, KeyEvent event) =>
          _handleTextFieldKeyEvent(event, submitOnEnter: true),
    );
    _shiftEnterfocusNode.addListener(_onShiftEnterFocusChange);
    _focusNode = FocusNode(
      onKeyEvent: (FocusNode node, KeyEvent event) =>
          _handleTextFieldKeyEvent(event, submitOnEnter: false),
    );
    _focusNode.addListener(_onFocusChange);
    widget.control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void dispose() {
    _controller.removeListener(_handleControllerChange);
    _controller.dispose();
    _shiftEnterfocusNode.removeListener(_onShiftEnterFocusChange);
    _shiftEnterfocusNode.dispose();
    _focusNode.removeListener(_onFocusChange);
    widget.control.removeInvokeMethodListener(_invokeMethod);
    _focusNode.dispose();
    super.dispose();
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("CupertinoTextField.$name($args)");
    switch (name) {
      case "focus":
        _focusNode.requestFocus();
      default:
        throw Exception("Unknown CupertinoTextField method: $name");
    }
  }

  void _onShiftEnterFocusChange() {
    setState(() {
      _focused = _shiftEnterfocusNode.hasFocus;
    });
    widget.control
        .triggerEvent(_shiftEnterfocusNode.hasFocus ? "focus" : "blur");
  }

  void _onFocusChange() {
    setState(() {
      _focused = _focusNode.hasFocus;
    });
    widget.control.triggerEvent(_focusNode.hasFocus ? "focus" : "blur");
  }

  void _handleControllerChange() {
    final selection = _controller.selection;
    if (_selection == selection) return;

    _selection = selection;

    if (!selection.isValid ||
        !widget.control.getBool("on_selection_change", false)!) {
      return;
    }

    widget.control.updateProperties({"selection": selection.toMap()});
    widget.control.triggerEvent("selection_change", {
      "selected_text":
          _controller.text.substring(selection.start, selection.end),
      "selection": selection.toMap()
    });
  }

  @override
  Widget build(BuildContext context) {
    debugPrint("CupertinoTextField build: ${widget.control.id}");

    var autofocus = widget.control.getBool("autofocus", false)!;
    var value = widget.control.getString("value", "")!;
    if (_value != value) {
      _value = value;
      _controller.value = TextEditingValue(
        text: value,
        // preserve cursor position at the end
        selection: TextSelection.collapsed(offset: value.length),
      );
      _selection = _controller.selection;
    }

    var shiftEnter = widget.control.getBool("shift_enter", false)!;
    var multiline = widget.control.getBool("multiline", false)! || shiftEnter;
    var minLines = widget.control.getInt("min_lines", 1)!;
    var maxLines = widget.control.getInt("max_lines", multiline ? null : 1);

    var readOnly = widget.control.getBool("read_only", false)!;
    var password = widget.control.getBool("password", false)!;
    var onChange = widget.control.getBool("on_change", false)!;

    var cursorColor = widget.control.getColor("cursor_color", context);
    var selectionColor = widget.control.getColor("selection_color", context);

    var maxLength = widget.control.getInt("max_length");

    var textSize = widget.control.getDouble("text_size");

    var color = widget.control.getColor("color", context);
    var focusedColor = widget.control.getColor("focused_color", context);

    var textStyle =
        widget.control.getTextStyle("text_style", Theme.of(context));
    if (textSize != null || color != null || focusedColor != null) {
      textStyle = (textStyle ?? const TextStyle()).copyWith(
          fontSize: textSize, color: _focused ? focusedColor ?? color : color);
    }

    List<TextInputFormatter> inputFormatters = [];
    var inputFilter = widget.control.getTextInputFormatter("input_filter");
    if (inputFilter != null) {
      inputFormatters.add(inputFilter);
    }
    var textCapitalization = widget.control
        .getTextCapitalization("capitalization", TextCapitalization.none)!;
    if (textCapitalization != TextCapitalization.none) {
      inputFormatters.add(TextCapitalizationFormatter(textCapitalization));
    }
    var textAlign = widget.control.getTextAlign("text_align", TextAlign.start)!;
    var textVerticalAlign = widget.control.getDouble("text_vertical_align");
    var rtl = widget.control.getBool("rtl", false)!;
    var autocorrect = widget.control.getBool("autocorrect", true)!;
    FocusNode focusNode = shiftEnter ? _shiftEnterfocusNode : _focusNode;

    var focusValue = widget.control.getString("focus");
    var blurValue = widget.control.getString("blur");
    if (focusValue != null && focusValue != _lastFocusValue) {
      _lastFocusValue = focusValue;
      focusNode.requestFocus();
    }
    if (blurValue != null && blurValue != _lastBlurValue) {
      _lastBlurValue = blurValue;
      _focusNode.unfocus();
    }

    var selection = widget.control.getTextSelection("selection",
        minOffset: 0, maxOffset: _controller.text.length);
    if (selection != null && selection != _controller.selection) {
      _controller.selection = selection;
      _selection = selection;
    }

    var borderRadius = widget.control.getBorderRadius("border_radius");

    BoxBorder? border;
    var borderWidth = widget.control.getDouble("border_width", 1.0)!;
    var borderColor = widget.control.getColor("border_color", context) ??
        const Color(0xFF000000);

    try {
      border = widget.control.getBorder("border", Theme.of(context));
      // adaptive TextField is being created
    } catch (e) {
      FormFieldInputBorder inputBorder = parseFormFieldInputBorder(
        widget.control.getString("border"),
        FormFieldInputBorder.outline,
      )!;

      if (inputBorder == FormFieldInputBorder.outline) {
        border = Border.all(color: borderColor, width: borderWidth);
      } else if (inputBorder == FormFieldInputBorder.underline) {
        border =
            Border(bottom: BorderSide(color: borderColor, width: borderWidth));
        borderRadius = BorderRadius.zero;
      }
    }

    var canRevealPassword =
        widget.control.getBool("can_reveal_password", false)!;

    Widget? revealPasswordIcon;
    if (password && canRevealPassword) {
      revealPasswordIcon = Padding(
        padding: const EdgeInsets.only(right: 15.0),
        child: GestureDetector(
            child: Icon(_revealPassword
                ? CupertinoIcons.eye_slash
                : CupertinoIcons.eye),
            onTap: () {
              setState(() {
                _revealPassword = !_revealPassword;
              });
            }),
      );
    }
    var fitParentSize = widget.control.getBool("fit_parent_size", false)!;
    var defaultDecoration = const CupertinoTextField().decoration;
    var gradient = widget.control.getGradient("gradient", Theme.of(context));
    var blendMode = widget.control.getBlendMode("blend_mode");
    var bgcolor = widget.control.getColor("bgcolor", context);
    var label = widget.control.get("label");
    String? labelStr;
    if (label is String) {
      labelStr = widget.control.getString("label");
    }

    Widget textField = CupertinoTextField(
        style: textStyle,
        textAlignVertical: textVerticalAlign != null
            ? TextAlignVertical(y: textVerticalAlign)
            : null,
        placeholder: widget.control.getString("placeholder_text") ?? labelStr,
        placeholderStyle:
            widget.control.getTextStyle("placeholder_style", Theme.of(context)) ??
                widget.control.getTextStyle("label_style", Theme.of(context)),
        // label_style for adaptive TextField
        autofocus: autofocus,
        enabled: !widget.control.disabled,
        onSubmitted: !multiline
            ? (String value) {
                widget.control.triggerEvent("submit", value);
              }
            : null,
        decoration: defaultDecoration?.copyWith(
            color: bgcolor,
            gradient: gradient,
            image: widget.control.getDecorationImage("image", context),
            backgroundBlendMode:
                bgcolor != null || gradient != null ? blendMode : null,
            border: border,
            borderRadius: borderRadius,
            boxShadow:
                widget.control.getBoxShadows("shadow", Theme.of(context))),
        cursorHeight: widget.control.getDouble("cursor_height"),
        showCursor: widget.control.getBool("show_cursor"),
        cursorWidth: widget.control.getDouble("cursor_width", 2.0)!,
        cursorRadius: widget.control
            .getRadius("cursor_radius", const Radius.circular(2.0))!,
        keyboardType: multiline
            ? TextInputType.multiline
            : widget.control
                .getTextInputType("keyboard_type", TextInputType.text)!,
        clearButtonSemanticLabel:
            widget.control.getString("clear_button_semantics_label"),
        autocorrect: autocorrect,
        enableSuggestions: widget.control.getBool("enable_suggestions", true)!,
        smartDashesType: widget.control.getBool("smart_dashes_type", true)!
            ? SmartDashesType.enabled
            : SmartDashesType.disabled,
        smartQuotesType: widget.control.getBool("smart_quotes_type", true)!
            ? SmartQuotesType.enabled
            : SmartQuotesType.disabled,
        suffixMode: widget.control.getOverlayVisibilityMode(
            "suffix_visibility_mode", OverlayVisibilityMode.always)!,
        prefixMode: widget.control.getOverlayVisibilityMode(
            "prefix_visibility_mode", OverlayVisibilityMode.always)!,
        textAlign: textAlign,
        minLines: fitParentSize ? null : minLines,
        maxLines: fitParentSize ? null : maxLines,
        maxLength: maxLength,
        prefix: widget.control.buildTextOrWidget("prefix"),
        suffix:
            revealPasswordIcon ?? widget.control.buildTextOrWidget("suffix"),
        readOnly: readOnly,
        textDirection: rtl ? TextDirection.rtl : null,
        inputFormatters: inputFormatters.isNotEmpty ? inputFormatters : null,
        obscureText: password && !_revealPassword,
        padding:
            widget.control.getPadding("padding", const EdgeInsets.all(7.0))!,
        stylusHandwritingEnabled:
            widget.control.getBool("enable_stylus_handwriting", true)!,
        scrollPadding: widget.control
            .getPadding("scroll_padding", const EdgeInsets.all(20.0))!,
        obscuringCharacter:
            widget.control.getString("obscuring_character", '•')!,
        cursorOpacityAnimates:
            widget.control.getBool("animate_cursor_opacity", isIOSMobile())!,
        expands: fitParentSize,
        enableIMEPersonalizedLearning:
            widget.control.getBool("enable_ime_personalized_learning", true)!,
        clipBehavior:
            widget.control.getClipBehavior("clip_behavior", Clip.hardEdge)!,
        cursorColor: cursorColor,
        autofillHints: parseAutofillHints(widget.control.get("autofill_hints")),
        keyboardAppearance: widget.control.getBrightness("keyboard_brightness"),
        enableInteractiveSelection:
            widget.control.getBool("enable_interactive_selection"),
        clearButtonMode: widget.control.getOverlayVisibilityMode("clear_button_visibility_mode", OverlayVisibilityMode.never)!,
        strutStyle: widget.control.getStrutStyle("strut_style"),
        onTap: () => widget.control.triggerEvent("click"),
        controller: _controller,
        focusNode: focusNode,
        onTapOutside: widget.control.getBool("on_tap_outside", false)!
            ? (PointerDownEvent? event) {
                widget.control.triggerEvent("tap_outside");
              }
            : null,
        onChanged: (String value) {
          _value = value;
          widget.control.updateProperties({"value": value});
          if (onChange) {
            widget.control.triggerEvent("change", value);
          }
        });

    if (cursorColor != null || selectionColor != null) {
      textField = TextSelectionTheme(
          data: TextSelectionTheme.of(context).copyWith(
              cursorColor: cursorColor, selectionColor: selectionColor),
          child: textField);
    }

    if (widget.control.getExpand("expand", 0)! > 0) {
      return LayoutControl(control: widget.control, child: textField);
    } else {
      double? width = widget.control.getDouble("width");

      return LayoutControl(
        control: widget.control,
        child: width == null
            ? ConstrainedBox(
                constraints: const BoxConstraints.tightFor(width: 300),
                child: textField)
            : textField,
      );
    }
  }
}

File: sdk/python/packages/flet-core/src/flet_core/cupertino_textfield.py
from enum import Enum
from typing import Any, List, Optional, Union

from flet_core.autofill_group import AutofillHint
from flet_core.border import Border
from flet_core.box import BoxShadow, DecorationImage, BoxShape
from flet_core.control import Control, OptionalNumber
from flet_core.gradients import Gradient
from flet_core.ref import Ref
from flet_core.text_style import TextStyle
from flet_core.textfield import InputFilter, KeyboardType, TextCapitalization, TextField
from flet_core.tooltip import TooltipValue
from flet_core.types import (
    AnimationValue,
    BlendMode,
    BorderRadiusValue,
    OffsetValue,
    PaddingValue,
    ResponsiveNumber,
    RotateValue,
    ScaleValue,
    TextAlign,
    OptionalControlEventCallable,
)


class VisibilityMode(Enum):
    NEVER = "never"
    EDITING = "editing"
    NOT_EDITING = "notEditing"
    ALWAYS = "always"


class CupertinoTextField(TextField):
    """
    An iOS-style text field.

    -----

    Online docs: https://flet.dev/docs/controls/cupertinotextfield
    """

    def __init__(
        self,
        placeholder_text: Optional[str] = None,
        value: Optional[str] = None,
        placeholder_style: Optional[TextStyle] = None,
        gradient: Optional[Gradient] = None,
        blend_mode: Optional[BlendMode] = None,
        shadow: Union[None, BoxShadow, List[BoxShadow]] = None,
        prefix_visibility_mode: Optional[VisibilityMode] = None,
        suffix_visibility_mode: Optional[VisibilityMode] = None,
        clear_button_semantics_label: Optional[str] = None,
        image: Optional[DecorationImage] = None,
        enable_scribble: Optional[bool] = None,
        padding: PaddingValue = None,
        scroll_padding: PaddingValue = None,
        obscuring_character: Optional[str] = None,
        on_click: OptionalControlEventCallable = None,
        #
        # TextField
        #
        keyboard_type: Optional[KeyboardType] = None,
        rtl: Optional[bool] = None,
        multiline: Optional[bool] = None,
        min_lines: Optional[int] = None,
        max_lines: Optional[int] = None,
        max_length: Optional[int] = None,
        password: Optional[bool] = None,
        can_reveal_password: Optional[bool] = None,
        read_only: Optional[bool] = None,
        shift_enter: Optional[bool] = None,
        text_align: Optional[TextAlign] = None,
        autofocus: Optional[bool] = None,
        capitalization: Optional[TextCapitalization] = None,
        autocorrect: Optional[bool] = None,
        enable_suggestions: Optional[bool] = None,
        smart_dashes_type: Optional[bool] = None,
        smart_quotes_type: Optional[bool] = None,
        cursor_color: Optional[str] = None,
        cursor_width: OptionalNumber = None,
        cursor_height: OptionalNumber = None,
        cursor_radius: OptionalNumber = None,
        show_cursor: Optional[bool] = None,
        selection_color: Optional[str] = None,
        input_filter: Optional[InputFilter] = None,
        autofill_hints: Union[None, AutofillHint, List[AutofillHint]] = None,
        on_change=None,
        on_submit=None,
        on_focus=None,
        on_blur=None,
        #
        # FormField
        #
        text_size: OptionalNumber = None,
        text_style: Optional[TextStyle] = None,
        border: Optional[Border] = None,
        color: Optional[str] = None,
        bgcolor: Optional[str] = None,
        border_radius: BorderRadiusValue = None,
        focused_color: Optional[str] = None,
        focused_bgcolor: Optional[str] = None,
        focused_border_width: OptionalNumber = None,
        focused_border_color: Optional[str] = None,
        content_padding: PaddingValue = None,
        dense: Optional[bool] = None,
        filled: Optional[bool] = None,
        prefix: Optional[Control] = None,
        suffix: Optional[Control] = None,
        #
        # ConstrainedControl
        #
        ref: Optional[Ref] = None,
        key: Optional[str] = None,
        width: OptionalNumber = None,
        height: OptionalNumber = None,
        expand: Union[None, bool, int] = None,
        expand_loose: Optional[bool] = None,
        col: Optional[ResponsiveNumber] = None,
        opacity: OptionalNumber = None,
        rotate: RotateValue = None,
        scale: ScaleValue = None,
        offset: OffsetValue = None,
        aspect_ratio: OptionalNumber = None,
        animate_opacity: AnimationValue = None,
        animate_size: AnimationValue = None,
        animate_position: AnimationValue = None,
        animate_rotation: AnimationValue = None,
        animate_scale: AnimationValue = None,
        animate_offset: AnimationValue = None,
        on_animation_end: OptionalControlEventCallable = None,
        tooltip: TooltipValue = None,
        visible: Optional[bool] = None,
        disabled: Optional[bool] = None,
        data: Any = None,
    ):
        TextField.__init__(
            self,
            ref=ref,
            key=key,
            width=width,
            height=height,
            expand=expand,
            expand_loose=expand_loose,
            col=col,
            opacity=opacity,
            rotate=rotate,
            scale=scale,
            offset=offset,
            aspect_ratio=aspect_ratio,
            animate_opacity=animate_opacity,
            animate_size=animate_size,
            animate_position=animate_position,
            animate_rotation=animate_rotation,
            animate_scale=animate_scale,
            animate_offset=animate_offset,
            on_animation_end=on_animation_end,
            tooltip=tooltip,
            visible=visible,
            disabled=disabled,
            data=data,
            #
            # FormField
            #
            text_size=text_size,
            text_style=text_style,
            color=color,
            bgcolor=bgcolor,
            border_radius=border_radius,
            focused_color=focused_color,
            focused_bgcolor=focused_bgcolor,
            focused_border_width=focused_border_width,
            focused_border_color=focused_border_color,
            content_padding=content_padding,
            dense=dense,
            filled=filled,
            prefix=prefix,
            suffix=suffix,
            #
            # TextField
            #
            value=value,
            keyboard_type=keyboard_type,
            rtl=rtl,
            multiline=multiline,
            min_lines=min_lines,
            max_lines=max_lines,
            max_length=max_length,
            password=password,
            can_reveal_password=can_reveal_password,
            read_only=read_only,
            shift_enter=shift_enter,
            text_align=text_align,
            autofocus=autofocus,
            capitalization=capitalization,
            autocorrect=autocorrect,
            enable_suggestions=enable_suggestions,
            smart_dashes_type=smart_dashes_type,
            smart_quotes_type=smart_quotes_type,
            cursor_color=cursor_color,
            cursor_width=cursor_width,
            cursor_height=cursor_height,
            cursor_radius=cursor_radius,
            show_cursor=show_cursor,
            selection_color=selection_color,
            input_filter=input_filter,
            autofill_hints=autofill_hints,
            on_change=on_change,
            on_submit=on_submit,
            on_focus=on_focus,
            on_blur=on_blur,
        )

        self.placeholder_text = placeholder_text
        self.placeholder_style = placeholder_style
        self.gradient = gradient
        self.blend_mode = blend_mode
        self.shadow = shadow
        self.suffix_visibility_mode = suffix_visibility_mode
        self.prefix_visibility_mode = prefix_visibility_mode
        self.clear_button_semantics_label = clear_button_semantics_label
        self.border = border
        self.image = image
        self.enable_scribble = enable_scribble
        self.padding = padding
        self.scroll_padding = scroll_padding
        self.obscuring_character = obscuring_character
        self.on_click = on_click

    def _get_control_name(self):
        return "cupertinotextfield"

    def before_update(self):
        super().before_update()
        self._set_attr_json("gradient", self.__gradient)
        self._set_attr_json("shadow", self.__shadow if self.__shadow else None)
        self._set_attr_json("placeholderStyle", self.__placeholder_style)
        self._set_attr_json("border", self.__border)
        self._set_attr_json("image", self.__image)
        self._set_attr_json("padding", self.__padding)
        self._set_attr_json("scrollPadding", self.__scroll_padding)

    # placeholder_text
    @property
    def placeholder_text(self):
        return self._get_attr("placeholderText")

    @placeholder_text.setter
    def placeholder_text(self, value):
        self._set_attr("placeholderText", value)

    # placeholder_style
    @property
    def placeholder_style(self):
        return self.__placeholder_style

    @placeholder_style.setter
    def placeholder_style(self, value: Optional[TextStyle]):
        self.__placeholder_style = value

    # gradient
    @property
    def gradient(self) -> Optional[Gradient]:
        return self.__gradient

    @gradient.setter
    def gradient(self, value: Optional[Gradient]):
        self.__gradient = value

    # blend_mode
    @property
    def blend_mode(self) -> Optional[BlendMode]:
        return self.__blend_mode

    @blend_mode.setter
    def blend_mode(self, value: Optional[BlendMode]):
        self.__blend_mode = value
        self._set_enum_attr("blendMode", value, BlendMode)

    # shadow
    @property
    def shadow(self) -> Union[None, BoxShadow, List[BoxShadow]]:
        return self.__shadow

    @shadow.setter
    def shadow(self, value: Union[None, BoxShadow, List[BoxShadow]]):
        self.__shadow = value if value is not None else []

    # image
    @property
    def image(self) -> Optional[DecorationImage]:
        return self.__image

    @image.setter
    def image(self, value: Optional[DecorationImage]):
        self.__image = value

    # shape
    @property
    def shape(self) -> Optional[BoxShape]:
        return self.__shape

    @shape.setter
    def shape(self, value: Optional[BoxShape]):
        self.__shape = value

    # suffix_visibility_mode
    @property
    def suffix_visibility_mode(self) -> Optional[VisibilityMode]:
        return self.__suffix_visibility_mode

    @suffix_visibility_mode.setter
    def suffix_visibility_mode(self, value: Optional[VisibilityMode]):
        self.__suffix_visibility_mode = value
        self._set_enum_attr("suffixVisibilityMode", value, VisibilityMode)

    # prefix_visibility_mode
    @property
    def prefix_visibility_mode(self) -> Optional[VisibilityMode]:
        return self.__prefix_visibility_mode

    @prefix_visibility_mode.setter
    def prefix_visibility_mode(self, value: Optional[VisibilityMode]):
        self.__prefix_visibility_mode = value
        self._set_enum_attr("prefixVisibilityMode", value, VisibilityMode)

    # clear_button_semantics_label
    @property
    def clear_button_semantics_label(self) -> Optional[str]:
        return self._get_attr("clearButtonSemanticsLabel")

    @clear_button_semantics_label.setter
    def clear_button_semantics_label(self, value: Optional[str]):
        self._set_attr("clearButtonSemanticsLabel", value)

    # obscuring_character
    @property
    def obscuring_character(self) -> Optional[str]:
        return self._get_attr("obscuringCharacter")

    @obscuring_character.setter
    def obscuring_character(self, value: Optional[str]):
        self._set_attr("obscuringCharacter", value)

    # enable_scribble
    @property
    def enable_scribble(self) -> bool:
        return self._get_attr("enableScribble", data_type="bool", def_value=True)

    @enable_scribble.setter
    def enable_scribble(self, value: Optional[bool]):
        self._set_attr("enableScribble", value)

    # padding
    @property
    def padding(self) -> PaddingValue:
        return self.__padding

    @padding.setter
    def padding(self, value: PaddingValue):
        self.__padding = value

    # scroll_padding
    @property
    def scroll_padding(self) -> PaddingValue:
        return self.__scroll_padding

    @scroll_padding.setter
    def scroll_padding(self, value: PaddingValue):
        self.__scroll_padding = value

    # on_click
    @property
    def on_click(self):
        return self._get_event_handler("click")

    @on_click.setter
    def on_click(self, handler: OptionalControlEventCallable):
        self._add_event_handler("click", handler)

    # border
    @property
    def border(self) -> Optional[Border]:
        return self.__border

    @border.setter
    def border(self, value: Optional[Border]):
        self.__border = value
